src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +30,109 @@ public static class Renamer
                 newName, optionSet, cancellationToken);
         }
 
+        /// <summary>
+        /// Applies changes to the given document so that the folders, name, and path from <paramref name="newDocumentInfo"/>
+        /// are used. Will also perform symbol changes to the code in the document, attempting to sync type and namespace
+        /// to the new information.
+        /// </summary>
+        /// <returns><see cref="RenameDocumentResult"/> with the modified solution and indication of any symbol changes made</returns>
+        public static async Task<RenameDocumentResult> RenameDocumentAsync(
+            Document document, DocumentInfo newDocumentInfo, OptionSet optionSet, CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            var newName = newDocumentInfo.Name;
+            var shouldRenameType = !string.Equals(document.Name, newName, StringComparison.OrdinalIgnoreCase);
+            var shouldMoveNamespace = !document.Folders.SequenceEqual(newDocumentInfo.Folders);
+
+            var solution = document.Project.Solution
+                .WithDocumentName(document.Id, newDocumentInfo.Name)
+                .WithDocumentFilePath(document.Id, newDocumentInfo.FilePath)
+                .WithDocumentFolders(document.Id, newDocumentInfo.Folders);
+
+            if (!(shouldRenameType || shouldMoveNamespace))
+            {
```

Comment:
open quesiton: if they're not renaming the type or the namespace... why are they calling this method in the first place?  should we consider that a bug in in the caller that we inform them about through argument validation?

Reply:
I don't think we should require the caller to know the contents of the document prior to calling this API, but should allow for changes if needed. This should be callable even if no changes other than DocumentInfo are required

Reply:
@CyrusNajmabadi: we're generally reserving the right to keep adding more things in the future. For example this will eventually update file headers, and other things you can imagine down road. So to a certain extent we don't want people trying to have their own "is this needed" logic -- they should just do it.

That said, @ryzngard did we lose the ability here to distinguish between whether there's even something to do? Like does the RenameDocumentResult need to have the true/false (or other structure as discussed) and then you have the separate GetNewSolutionAsync() that does the potentially expensive rename? Because the project system doesn't want us spending a minute computing an expensive rename before asking the user...